File: ctci/ctci-queue-using-two-stacks.pl
#!/usr/bin/env perl
use strict;

=head1 Hacker Rank ctci-queue-using-two-stacks

    mitch@mitchjacksontech.com
    https://www.hackerrank.com/challenges/ccti-queue-using-two-stacks

=head1 Comments

    Since OO is very slow on hackerrank, because hr refuses to
    allow perl's OO libraries, let's implement as a simple array

=cut

my $first_line = 1;
my @q;

while (<>) {
    if ($first_line) { $first_line = 0; next; }
    chomp;
    my ($cmd,$val) = split / /,$_;

    if    ( $cmd == 1 ) { push @q, $val; } # Enqueue
    elsif ( $cmd == 2 ) { shift(@q); }     # Dequeue
    elsif ( $cmd == 3 ) { print scalar(@q) ? $q[0] : undef; print "\n"; } # peek
}

File: ctci/ctci-queue-using-two-stacks.py
#!/usr/bin/env python3
from sys import stdin

# Hacker Rank Queues: A Tale of Two Stacks
#  mitch@mitchjacksontech.com
#  https://www.hackerrank.com/challenges/ctci-queue-using-two-stacks
#
# The solution calls for implementing a queue object interacting with it.

class my_queue(object):
    def __init__(self):
        self.stack = []

    def peek(self):
        if len(self.stack):
            return self.stack[0]
        return None

    def pop(self):
        if len(self.stack):
            return self.stack.pop(0)
        return None

    def put(self, val):
        return self.stack.append(val)

queue = my_queue();

for line in stdin:
    vals = list(map(int,line.split()))

    # Enqueue
    if vals[0] == 1:
        queue.put(vals[1])

    # Dequeue
    elif vals[0] == 2:
        queue.pop()

    # Peek
    elif vals[0] == 3:
        print(queue.peek())
